# Exercises I
Create a notebook or python script named advanced_dataframes to do your work in for these exercises.

1. Run python -m pip install pymysql from your terminal to install the mysql client (any folder is fine)
2. cd into your exercises folder for this module and run echo env.py >> .gitignore
3. Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

4. Use your function to obtain a connection to the employees database.

5. Once you have successfully run a query:

- a. Intentionally make a typo in the database url. What kind of error message do you see?

- b. Intentionally make an error in your SQL query. What does the error message look like?

6. Read the employees and titles tables into two separate DataFrames.

7. How many rows and columns do you have in each DataFrame? Is that what you expected?

8. Display the summary statistics for each DataFrame.

9. How many unique titles are in the titles DataFrame?

10. What is the oldest date in the to_date column?

11. What is the most recent date in the to_date column?

In [1]:
import pandas as pd
import numpy as np

### 3. Create a function named get_db_url. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

In [2]:
def get_db_url(db_name):
    from env import host, user, password
    return f'mysql+pymysql://{user}:{password}@{host}/{db_name}'
    
# def get_db_url(user,password,host,db_name):
#     return f'mysql+pymysql://{user}:{password}@{host}/{db_name}'


# url = get_db_url(user,password,host,'employees')

### 4. Use your function to obtain a connection to the employees database.



In [3]:
employees_url = get_db_url('employees')

# from env import host, user, password
# url = get_db_url(user,password,host,'employees')

In [4]:
employees_query = 'Select * from employees'

In [5]:
df = pd.read_sql(employees_query, employees_url)
df

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

### 5. Once you have successfully run a query:

a. Intentionally make a typo in the database url. What kind of error message do you see?

b. Intentionally make an error in your SQL query. What does the error message look like?

### THESE CELLS ARE COMMENTED OUT SO I CAN RESTART AND RUN ALL

In [6]:
# def get_bad_url(db_name):
#     from env import host, user, password
#     return f'mysql+pymysql://{user}:{password}@fdas{host}/{db_name}'

# bad_url = get_bad_url('employees')
# bad_query = 'Select all from employees'

# errorA = pd.read_sql(query, bad_url)
# errorA

In [7]:
# errorB = pd.read_sql(bad_query, url)
# errorB

### 6. Read the employees and titles tables into two separate DataFrames.



In [8]:
employees = pd.read_sql(employees_query, employees_url)
employees

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [9]:
query_titles = 'Select * from titles'

In [10]:
titles = pd.read_sql(query_titles, employees_url)
titles

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

### 7. How many rows and columns do you have in each DataFrame? Is that what you expected?

In [11]:
employees.shape[0]

300024

In [12]:
titles.shape[0]

443308

Employees = 300024 rows, Titles = 443308

### 8.  Display the summary statistics for each DataFrame.



In [13]:
employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


In [14]:
employees.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [15]:
titles.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


In [16]:
titles.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

### 9. How many unique titles are in the titles DataFrame?



In [17]:
len(titles.title.value_counts())

7

### 10. What is the oldest date in the to_date column?



In [18]:
titles.to_date.min()

datetime.date(1985, 3, 1)

### 11. What is the most recent date in the to_date column?



In [19]:
titles_date_query = """SELECT * From titles WHERE to_date != '9999-01-01'"""
titles_recent_date_df = pd.read_sql(titles_date_query, employees_url)
titles_recent_date_df.to_date

0         1995-12-01
1         1996-09-12
2         1996-02-11
3         2000-07-31
4         1990-02-18
             ...    
203179    1992-05-09
203180    1993-10-27
203181    2002-05-13
203182    1992-08-29
203183    1998-12-27
Name: to_date, Length: 203184, dtype: object

In [20]:
titles.to_date.max()
#or filtered to exclude 9999-01-01
titles_recent_date_df.to_date.max()

titles[titles.to_date != titles.to_date.max()].to_date.max()

datetime.date(2002, 8, 1)

# Exercises II

1. Copy the users and roles DataFrames from the examples above.

2. What is the result of using a right join on the DataFrames?

3. What is the result of using an outer join on the DataFrames?

4. What happens if you drop the foreign keys from the DataFrames and try to merge them?

5. Load the mpg dataset from PyDataset.

6. Output and read the documentation for the mpg dataset.

7. How many rows and columns are in the dataset?

8. Check out your column names and perform any cleanup you may want on them.

9. Display the summary statistics for the dataset.

10. How many different manufacturers are there?

11. How many different models are there?

12. Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

13. Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

14. Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.

16. Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?

17. Do automatic or manual cars have better miles per gallon?




### 1. Copy the users and roles DataFrames from the examples above.

In [21]:
# Create the users DataFrame.

users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [22]:
# Create the roles DataFrame

roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

### 2. What is the result of using a right join on the DataFrames?



In [23]:
users.merge(roles, left_on='role_id', right_on='id', how='right', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0     1      admin        both
1   2.0    joe      2.0     2     author        both
2   3.0  sally      3.0     3   reviewer        both
3   4.0   adam      3.0     3   reviewer        both
4   NaN    NaN      NaN     4  commenter  right_only

- matched column names are added _x or _y
- all rows from right df are kept
- matching rows from left df are combined
- unmatched are excluded
- unmatched from right are included and added NaN for missing values

### 3. What is the result of using an outer join on the DataFrames?



In [24]:
users.merge(roles, left_on='role_id', right_on='id', how='outer', indicator=True)

id_x name_x  role_id  id_y     name_y      _merge
0   1.0    bob      1.0   1.0      admin        both
1   2.0    joe      2.0   2.0     author        both
2   3.0  sally      3.0   3.0   reviewer        both
3   4.0   adam      3.0   3.0   reviewer        both
4   5.0   jane      NaN   NaN        NaN   left_only
5   6.0   mike      NaN   NaN        NaN   left_only
6   NaN    NaN      NaN   4.0  commenter  right_only

- matching column names are added _x or _y
- all rows kept
- rows that match are combined
- unmatched are added and NaN is added for missing values

### 4. What happens if you drop the foreign keys from the DataFrames and try to merge them?



In [25]:
users.merge(roles, how='inner')

Empty DataFrame
Columns: [id, name, role_id]
Index: []

- foreign keys are needed to match
- without them, the result is empty

### 5. Load the mpg dataset from PyDataset.



In [26]:
from pydataset import data

In [27]:
mpg = data('mpg')

### 6. Output and read the documentation for the mpg dataset.



In [28]:
data('mpg', show_doc = True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




### 7. How many rows and columns are in the dataset?



In [29]:
print(f'There are {mpg.shape[0]} rows and {mpg.shape[1]} columns.')

There are 234 rows and 11 columns.


### 8. Check out your column names and perform any cleanup you may want on them.



In [30]:
mpg.rename(columns={'hwy':'highway', 'cty':'city', 
                    'cyl':'cylinders', 'drv':'drivetrain',
                   'displ':'displacement', 'trans':'transmission', 'fl':'fuel'}, inplace = True)
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drivetrain  city  highway fuel    class  
1            f    18       29    p  compact  
2            f    21       29    p  compact  
3            f    20       31    p  compact  
4            f    21       30    p  compact  
5            f    16       26    p  compact  
..         ...   ...      ...  ...      ...  
230          f    19       28    p  midsize  
231          f    21       29    p  midsize  
232          f    16       26    p  midsize  
233          f    18       26    p  midsize  
234          f    17       26    p  midsize  

[234 rows x 11 columns]

### 9. Display the summary statistics for the dataset.



In [31]:
mpg.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 234 entries, 1 to 234
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   manufacturer  234 non-null    object 
 1   model         234 non-null    object 
 2   displacement  234 non-null    float64
 3   year          234 non-null    int64  
 4   cylinders     234 non-null    int64  
 5   transmission  234 non-null    object 
 6   drivetrain    234 non-null    object 
 7   city          234 non-null    int64  
 8   highway       234 non-null    int64  
 9   fuel          234 non-null    object 
 10  class         234 non-null    object 
dtypes: float64(1), int64(4), object(6)
memory usage: 21.9+ KB


In [32]:
mpg.describe()

displacement         year   cylinders        city     highway
count    234.000000   234.000000  234.000000  234.000000  234.000000
mean       3.471795  2003.500000    5.888889   16.858974   23.440171
std        1.291959     4.509646    1.611534    4.255946    5.954643
min        1.600000  1999.000000    4.000000    9.000000   12.000000
25%        2.400000  1999.000000    4.000000   14.000000   18.000000
50%        3.300000  2003.500000    6.000000   17.000000   24.000000
75%        4.600000  2008.000000    8.000000   19.000000   27.000000
max        7.000000  2008.000000    8.000000   35.000000   44.000000

### 10. How many different manufacturers are there?



In [33]:
len(mpg.manufacturer.value_counts())

15

### 11. How many different models are there?



In [34]:
len(mpg.model.value_counts())

38

### 12. Create a column named mileage_difference like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.



In [35]:
mpg['mileage_difference'] = mpg.highway - mpg.city
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drivetrain  city  highway fuel    class  mileage_difference  
1            f    18       29    p  compact                  11  
2            f    21       29    p  compact                   8  
3            f    20       31    p  compact                  11  
4            f    21       30    p  compact                   9  
5            f    16       26    p  compact                  10  
..         ...   ...      ...  ...      ...                 ...  
230          f    19       28    p  midsize                   9  
231          f    21       29    p  midsize                   8  
232          f    16       26    p  midsize                  10  
233          f    18       26    p  midsize                   8  
234          f    17       26    p  midsize                   9  

[234 rows x 12 columns]

### 13. Create a column named average_mileage like you did in the DataFrames exercises; this is the mean of the city and highway mileage.



In [36]:
mpg['average_mileage'] = (mpg.city + mpg.highway)/2
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drivetrain  city  highway fuel    class  mileage_difference  \
1            f    18       29    p  compact                  11   
2            f    21       29    p  compact                   8   
3            f    20       31    p  compact                  11   
4            f    21       30    p  compact                   9   
5            f    16       26    p  compact                  10   
..         ...   ...      ...  ...      ...                 ...   
230          f    19       28    p  midsize                   9   
231          f    21       29    p  midsize                   8   
232          f    16       26    p  midsize                  10   
233          f    18       26    p  midsize                   8   
234          f    17       26    p  midsize                   9   

     average_mileage  
1               23.5  
2               25.0  
3               25.5  
4               25.5  
5               21.0  
..               ...  
230             23.5  
231             25.0  
232             21.0  
233             22.0  
234             21.5  

[234 rows x 13 columns]

### 14. Create a new column on the mpg dataset named is_automatic that holds boolean values denoting whether the car has an automatic transmission.



In [37]:
mpg['is_automatic'] = np.where(mpg.transmission.str.startswith('auto'), True, False)
mpg

manufacturer   model  displacement  year  cylinders transmission  \
1           audi      a4           1.8  1999          4     auto(l5)   
2           audi      a4           1.8  1999          4   manual(m5)   
3           audi      a4           2.0  2008          4   manual(m6)   
4           audi      a4           2.0  2008          4     auto(av)   
5           audi      a4           2.8  1999          6     auto(l5)   
..           ...     ...           ...   ...        ...          ...   
230   volkswagen  passat           2.0  2008          4     auto(s6)   
231   volkswagen  passat           2.0  2008          4   manual(m6)   
232   volkswagen  passat           2.8  1999          6     auto(l5)   
233   volkswagen  passat           2.8  1999          6   manual(m5)   
234   volkswagen  passat           3.6  2008          6     auto(s6)   

    drivetrain  city  highway fuel    class  mileage_difference  \
1            f    18       29    p  compact                  11   
2            f    21       29    p  compact                   8   
3            f    20       31    p  compact                  11   
4            f    21       30    p  compact                   9   
5            f    16       26    p  compact                  10   
..         ...   ...      ...  ...      ...                 ...   
230          f    19       28    p  midsize                   9   
231          f    21       29    p  midsize                   8   
232          f    16       26    p  midsize                  10   
233          f    18       26    p  midsize                   8   
234          f    17       26    p  midsize                   9   

     average_mileage  is_automatic  
1               23.5          True  
2               25.0         False  
3               25.5         False  
4               25.5          True  
5               21.0          True  
..               ...           ...  
230             23.5          True  
231             25.0         False  
232             21.0          True  
233             22.0         False  
234             21.5          True  

[234 rows x 14 columns]

### 15. Using the mpg dataset, find out which which manufacturer has the best miles per gallon on average?



In [39]:
manufacturer_groups = mpg.groupby(['manufacturer'])
manufacturer_groups.average_mileage.mean().sort_values(ascending = False).head(1)

manufacturer
honda    28.5
Name: average_mileage, dtype: float64

### 16. Do automatic or manual cars have better miles per gallon?


In [61]:
transmission_groups = mpg.groupby(['is_automatic'])
transmission_groups.average_mileage.mean().sort_values(ascending = False)

is_automatic
False    22.227273
True     19.130573
Name: average_mileage, dtype: float64

# Exercises III

1. Use your get_db_url function to help you explore the data from the chipotle database.

2. What is the total price for each order?

3. What are the most popular 3 items?

4. Which item has produced the most revenue?

5. Join the employees and titles DataFrames together.

6. For each title, find the hire date of the employee that was hired most recently with that title.

7. Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)



### 1. Use your get_db_url function to help you explore the data from the chipotle database.

In [41]:
chipotle_url = get_db_url('chipotle')
chipotle_query = """
        SELECT * from orders
        """

In [42]:
chipotle = pd.read_sql(chipotle_query, chipotle_url)
chipotle.sample(10)

id  order_id  quantity                     item_name  \
4491  4492      1786         4             Canned Soft Drink   
4255  4256      1698         1                  Chicken Bowl   
891    892       368         1                Veggie Burrito   
494    495       207         1              Steak Soft Tacos   
4142  4143      1657         1             Canned Soft Drink   
422    423       184         1                  Chicken Bowl   
188    189        84         1               Chicken Burrito   
4203  4204      1678         1                  Chicken Bowl   
3249  3250      1301         1                  Chicken Bowl   
3662  3663      1464         1  Chips and Fresh Tomato Salsa   

                                                                           choice_description  \
4491                                                                                 [Sprite]   
4255                  [Tomatillo Green Chili Salsa, [Rice, Pinto Beans, Sour Cream, Lettuce]]   
891             [Fresh Tomato Salsa (Mild), [Black Beans, Rice, Cheese, Sour Cream, Lettuce]]   
494                                                [Fresh Tomato Salsa, [Sour Cream, Cheese]]   
4142                                                                                   [Coke]   
422                                               [Roasted Chili Corn Salsa, [Rice, Lettuce]]   
188                  [Roasted Chili Corn Salsa (Medium), [Rice, Cheese, Sour Cream, Lettuce]]   
4203            [Tomatillo Red Chili Salsa, [Rice, Black Beans, Cheese, Sour Cream, Lettuce]]   
3249  [Roasted Chili Corn Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
3662                                                                                      nan   

     item_price  
4491     $5.00   
4255     $8.75   
891      $8.49   
494      $9.25   
4142     $1.25   
422      $8.75   
188      $8.49   
4203     $8.75   
3249    $11.25   
3662     $2.95

In [43]:
# Convert item_price from str to float
chipotle['item_price'] = chipotle.item_price.str.replace('$','').astype(float)
chipotle

<ipython-input-43-297741027801>:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  chipotle['item_price'] = chipotle.item_price.str.replace('$','').astype(float)


id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

      item_price  
0           2.39  
1           3.39  
2           3.39  
3           2.39  
4          16.98  
...          ...  
4617       11.75  
4618       11.75  
4619       11.25  
4620        8.75  
4621        8.75  

[4622 rows x 6 columns]

In [44]:
# chipotle['order_total'] = 
chipotle.groupby('order_id').item_price.sum()
chipotle

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

      item_price  
0           2.39  
1           3.39  
2           3.39  
3           2.39  
4          16.98  
...          ...  
4617       11.75  
4618       11.75  
4619       11.25  
4620        8.75  
4621        8.75  

[4622 rows x 6 columns]

### 2. What is the total price for each order?



In [45]:
chipotle.groupby('order_id').item_price.sum()

order_id
1       11.56
2       16.98
3       12.67
4       21.00
5       13.70
        ...  
1830    23.00
1831    12.90
1832    13.20
1833    23.50
1834    28.75
Name: item_price, Length: 1834, dtype: float64

### 3. What are the most popular 3 items?



In [63]:
chipotle.groupby('item_name').quantity.sum().sort_values(ascending = False).head(3)

item_name
Chicken Bowl           761
Chicken Burrito        591
Chips and Guacamole    506
Name: quantity, dtype: int64

### 4. Which item has produced the most revenue?



In [47]:
chipotle.groupby('item_name').item_price.sum().sort_values().tail(1)

item_name
Chicken Bowl    7342.73
Name: item_price, dtype: float64

### 5. Join the employees and titles DataFrames together.



In [50]:
employees_titles = employees.merge(titles, left_on='emp_no', right_on='emp_no', how='inner')
employees_titles

emp_no  birth_date first_name last_name gender   hire_date  \
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
...        ...         ...        ...       ...    ...         ...   
443303  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443304  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443305  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443306  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443307  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30   

                  title   from_date     to_date  
0       Senior Engineer  1986-06-26  9999-01-01  
1                 Staff  1996-08-03  9999-01-01  
2       Senior Engineer  1995-12-03  9999-01-01  
3              Engineer  1986-12-01  1995-12-01  
4       Senior Engineer  1995-12-01  9999-01-01  
...                 ...         ...         ...  
443303         Engineer  1987-08-30  1992-08-29  
443304  Senior Engineer  1992-08-29  9999-01-01  
443305     Senior Staff  1998-12-27  9999-01-01  
443306            Staff  1993-12-27  1998-12-27  
443307         Engineer  1997-11-30  9999-01-01  

[443308 rows x 9 columns]

### 6. For each title, find the hire date of the employee that was hired most recently with that title.

In [51]:
employees_titles.groupby('title').hire_date.max()

title
Assistant Engineer    1999-12-24
Engineer              2000-01-28
Manager               1992-02-05
Senior Engineer       2000-01-01
Senior Staff          2000-01-13
Staff                 2000-01-12
Technique Leader      1999-12-31
Name: hire_date, dtype: object

### 7. Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

In [52]:
departments_query = 'Select * from departments'
departments = pd.read_sql(departments_query, employees_url)
departments

dept_no           dept_name
0    d009    Customer Service
1    d005         Development
2    d002             Finance
3    d003     Human Resources
4    d001           Marketing
5    d004          Production
6    d006  Quality Management
7    d008            Research
8    d007               Sales

In [53]:
dept_emp_query = 'Select * from dept_emp'
dept_emp = pd.read_sql(dept_emp_query, employees_url)
dept_emp

emp_no dept_no   from_date     to_date
0        10001    d005  1986-06-26  9999-01-01
1        10002    d007  1996-08-03  9999-01-01
2        10003    d004  1995-12-03  9999-01-01
3        10004    d004  1986-12-01  9999-01-01
4        10005    d003  1989-09-12  9999-01-01
...        ...     ...         ...         ...
331598  499995    d004  1997-06-02  9999-01-01
331599  499996    d004  1996-05-13  9999-01-01
331600  499997    d005  1987-08-30  9999-01-01
331601  499998    d002  1993-12-27  9999-01-01
331602  499999    d004  1997-11-30  9999-01-01

[331603 rows x 4 columns]

In [54]:
query_current_titles = 'Select * from titles where to_date > CURDATE()'
cur_titles = pd.read_sql(query_current_titles, employees_url)

### Solution with merge and pd.crosstab()

- Do not forget you can merge in the SQL query

In [55]:
# Merge titles to dept_emp
dept_emp_titles = cur_titles.merge(dept_emp, left_on='emp_no', right_on='emp_no', how='inner')
# Merge ^df to departments to add dept_names
departments_titles = departments.merge(dept_emp_titles, left_on='dept_no',right_on='dept_no',how='inner')

In [56]:
# Use dept_name as index, title for columns
pd.crosstab(departments_titles.dept_name, departments_titles.title)

title               Assistant Engineer  Engineer  Manager  Senior Engineer  \
dept_name                                                                    
Customer Service                    68       627        1             1790   
Development                       1833     15677        1            43364   
Finance                              0         0        1                0   
Human Resources                      0         0        1                0   
Marketing                            0         0        1                0   
Production                        1542     13325        1            37156   
Quality Management                 433      3744        1            10390   
Research                            77       830        1             2250   
Sales                                0         0        1                0   

title               Senior Staff  Staff  Technique Leader  
dept_name                                                  
Customer Service           12349   3902               241  
Development                 1085    315              6117  
Finance                    10650   3199                 0  
Human Resources            10843   3416                 0  
Marketing                  12371   3880                 0  
Production                  1123    349              5210  
Quality Management             0      0              1422  
Research                   10219   3206               321  
Sales                      32032   9967                 0

### how not to do it

In [57]:
# dept_emp_departments = dept_emp.merge(departments, left_on='dept_no',right_on='dept_no', how='left')
# pd.crosstab(titles.title, dept_emp_departments.dept_name)

In [58]:
# pd.crosstab(titles.title, dept_emp.dept_no)

### Solution by merge and groupby

In [59]:
## Solution with merge and groupby
dept_emp_titles = cur_titles.merge(dept_emp, left_on='emp_no', right_on='emp_no', how='inner')
departments_titles = departments.merge(dept_emp_titles, left_on='dept_no',right_on='dept_no',how='inner')
departments_titles.groupby('dept_name').title.value_counts()

dept_name           title             
Customer Service    Senior Staff          12349
                    Staff                  3902
                    Senior Engineer        1790
                    Engineer                627
                    Technique Leader        241
                    Assistant Engineer       68
                    Manager                   1
Development         Senior Engineer       43364
                    Engineer              15677
                    Technique Leader       6117
                    Assistant Engineer     1833
                    Senior Staff           1085
                    Staff                   315
                    Manager                   1
Finance             Senior Staff          10650
                    Staff                  3199
                    Manager                   1
Human Resources     Senior Staff          10843
                    Staff                  3416
                    Manager                   1
M